jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jmap.java

```diff
@@ -0,0 +1,159 @@
+	private static boolean histo;
+	private static boolean live;
+	@SuppressWarnings("nls")
+	private static String HELPTEXT = "jmap: obtain heap information about a Java process%n"
+			+ " Usage:%n"
+			+ "    jmap <option>* <vmid>%n"
+			+ "        <vmid>: Attach API VM ID as shown in jps or other Attach API-based tools%n"
+			+ "        <vmid>s are read from stdin if none are supplied as arguments%n"
+			+ "    -histo: print statistics about classes on the heap, including number of objects and aggregate size%n"
+			+ "    -histo:live : Print only live objects%n"
+			+ "    -J: supply arguments to the Java VM running jmap%n"
+			+ "NOTE: this utility may significantly affect the performance of the target VM.%n"
+			+ "At least one option must be selected.%n";
+
+	/**
+	 * Print a list of Java processes and information about them.
+	 * 
+	 * @param args Arguments to the application
+	 */
+	public static void main(String[] args) {
+
+		if (!parseArguments(args)) {
+			System.exit(1);
+		}
+		AttacherDiagnosticsProvider diagProvider = new AttacherDiagnosticsProvider();
+
+		String myId = AttachHandler.getVmId();
+		for (String vmid : vmids) {
+			if (vmid.equals(myId)) {
+				continue;
+			}
+			/* if the ID looks like a process ID, check if it is running */
+			if (vmid.matches("\\d+")) { //$NON-NLS-1$
+				long pid = Long.parseLong(vmid);
+				if (!IPC.processExists(pid)) {
+					System.out.println("No such process: " + vmid); //$NON-NLS-1$
+					continue;
+				}
+			}
+			try {
+				if (vmids.size() > 1) {
+					System.out.printf("Virtual machine: %s%n", vmid); //$NON-NLS-1$
+				}
+				diagProvider.attach(vmid);
+				if (histo) {
+					runAndPrintCommand(diagProvider, DiagnosticUtils.makeHeapHistoCommand(live));
+				}
+			} catch (Exception e) {
+				System.err.printf("Error getting data from %s", vmid); //$NON-NLS-1$
+				final String msg = e.getMessage();
+				if (null == msg) {
+					System.err.println();
+				} else {
+					if (msg.matches(IPC.INCOMPATIBLE_JAVA_VERSION)) {
+						System.err.println(": incompatible target JVM"); //$NON-NLS-1$
+					} else {
+						System.err.printf(": %s%n", msg); //$NON-NLS-1$
+					}
+				}
+				if (DiagnosticProperties.isDebug) {
+					e.printStackTrace();
+				}
+			} finally {
+				try {
+					diagProvider.detach();
+				} catch (IOException e) {
+					/* ignore */
+				}
+			}
+		}
+	}
+
+	private static void runAndPrintCommand(AttacherDiagnosticsProvider diagProvider, String cmd) throws IOException {
+		Properties props = diagProvider.executeDiagnosticCommand(cmd);
+		DiagnosticProperties.dumpPropertiesIfDebug("jmap result:", props); //$NON-NLS-1$
+		String responseString = new DiagnosticProperties(props).printStringResult();
+		System.out.print(responseString);
+	}
+
+	@SuppressWarnings("nls")
+	private static boolean parseArguments(String[] args) {
+		boolean okay = true;
+		vmids = new ArrayList<>();
+		histo = false;
+		live = true;
+		int optionsSelected = 0;
+		String errorMessage = "";
```

Comment:
We don't need both `okay` and `errorMessage`; the latter being null should be interpreted as OK.